pytest_playwright/pytest_playwright.py

```diff
@@ -419,3 +442,112 @@ def pytest_addoption(parser: Any) -> None:
         default=False,
         help="Whether to take a full page screenshot",
     )
+
+
+class ArtifactsRecorder:
+    def __init__(
+        self, pytestconfig: Any, request: pytest.FixtureRequest, playwright: Playwright
+    ) -> None:
+        self._request = request
+        self._pytestconfig = pytestconfig
+        self._playwright = playwright
+
+        self._open_contexts: BrowserContext = []
+        self._all_pages: List[Page] = []
+        self._traces: List[str] = []
+        self._tracing_option = pytestconfig.getoption("--tracing")
+        self._capture_trace = self._tracing_option in ["on", "retain-on-failure"]
+
+    def did_finish_test(self, failed: bool) -> None:
+        screenshot_option = self._pytestconfig.getoption("--screenshot")
+        capture_screenshot = screenshot_option == "on" or (
+            failed and screenshot_option == "only-on-failure"
+        )
+        if capture_screenshot:
+            for index, page in enumerate(self._all_pages):
+                human_readable_status = "failed" if failed else "finished"
+                screenshot_path = _build_artifact_test_folder(
+                    self._pytestconfig,
+                    self._request,
+                    f"test-{human_readable_status}-{index+1}.png",
+                )
+                try:
+                    page.screenshot(
+                        timeout=5000,
+                        path=screenshot_path,
+                        full_page=self._pytestconfig.getoption(
+                            "--full-page-screenshot"
+                        ),
+                    )
+                except Error:
+                    pass
+
+        # Close contexts which were not closed during the test (this will trigger Trace and Video generation)
+        while len(self._open_contexts) > 0:
```

Comment:
Instead of this, we can close in `new_context` teardown, and separately capture a screenshot in `will_close_context`.

Reply:
Addressed.